### PR TITLE
[data] New executor backend [3/n]--- Add basic operators impl

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Add the initial operator implementations.

This is split out from https://github.com/ray-project/ray/pull/30903

TODO:
- [x] Add unit test for AllToAllOp
- [x] Add unit test for MapOp
- [x] Add unit test for InputDataBuffer

### CODE REVIEW

File: python/ray/data/tests/test_operators.py
Comment (ericl) [inline python/ray/data/tests/test_operators.py:147]: Debating whether it's worth it to mock out the Ray API here to speed up these tests a bit. Maybe it's not that important since the bulk of the testing will be for StreamingExecutor, which we can write separate mocks for.

File: python/ray/data/_internal/execution/operators/input_data_buffer.py
Comment (stephanie-wang) [inline python/ray/data/_internal/execution/operators/input_data_buffer.py:21]:
```suggestion
            input_data: The list of bundles to output from this operator.
```

File: python/ray/data/_internal/execution/operators/input_data_buffer.py
Comment (stephanie-wang) [inline python/ray/data/_internal/execution/operators/input_data_buffer.py:17]: Hmm I thought the inputs to read tasks wouldn't always be Refs, like a list of filenames?

File: python/ray/data/_internal/execution/operators/map_operator.py
Comment (stephanie-wang) [inline python/ray/data/_internal/execution/operators/map_operator.py:85]: Hmm I wonder if we can keep the implementation details of compute strategy and ray remote args etc outside of the operators? It could be cleaner if we pass in the ray.remote Callable instead of the worker's Callable as the `transform_fn` but not sure if this will work so I'll leave it up to you.

File: python/ray/data/_internal/execution/operators/input_data_buffer.py
Comment (ericl) [reply python/ray/data/_internal/execution/operators/input_data_buffer.py:17]: Yeah, that would be true once we add the Read operator (which doesn't need to subclass from InputDataBuffer). Currently, in the legacy translation layer we ray.put() the read tasks, but this should go away in the future.

File: python/ray/data/_internal/execution/operators/map_operator.py
Comment (ericl) [reply python/ray/data/_internal/execution/operators/map_operator.py:85]: Yeah, I have a TODO on line 78 to clean this up in the future. I'm hoping the ComputeStrategy can turn into a simple dataclass once we migrate fully to the new backend. Right now, I avoided doing this refactoring to keep the changes self-contained.

About the callable, I think that's possible but it's probably also easier to do once we have the logical optimization layer in place (the optimizer could generate the ray.remote callable).

**Review by stephanie-wang — APPROVED** [review]
Looks great!

**Comment by ericl** [comment]
I'll hold this open until EOD for more comments.

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (c21) [inline python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:60]: shall we add type hint for `_block_bundle` as well? It would be great to add one-liner comment for what it is as well.

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (c21) [inline python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:76]: nit: `bundle: Optional[RefBundle]`

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (c21) [inline python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:164]: generator API / dynamic block splitting does not work with Ray client yet (https://github.com/ray-project/ray/issues/30458). We should guard the logic with `DatasetContext.get_current().block_splitting_enabled`.

Alternative is to not enable new execution backend if Ray client is used.

File: python/ray/data/_internal/execution/util.py
Comment (c21) [inline python/ray/data/_internal/execution/util.py:8]: shall we add comments for these utility methods, as they might be used across multiple places?

Also why are they named with leading `_`?
`_make_ref_bundles` -> `make_ref_bundles`?

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (c21) [inline python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:163]: Previously we have a nice warning for users in case blocks bundling reduces parallelism too much ([code](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/compute.py#L519-L527)). Is it possible to have a simliar warning here?

File: python/ray/data/_internal/execution/operators/map_operator.py
Comment (c21) [inline python/ray/data/_internal/execution/operators/map_operator.py:46]: Should we name it as `min_rows_per_fn_call`? `batch` is kind of confusing here, as this is neither user-facing batch in `map_batches`, nor zero-copy batch execution we shall introduce later.

Comment (ericl) [reply python/ray/data/_internal/execution/operators/map_operator.py:46]: Batch seems clearer to me: it basically is the same as the user facing batch size.

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (ericl) [reply python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:163]: This is a little trickier in the streaming execution model, maybe we can add it back later as needed.

File: python/ray/data/_internal/execution/operators/map_operator.py
Comment (ericl) [reply python/ray/data/_internal/execution/operators/map_operator.py:46]: I don't think so, the previous naming was very confusing for me. The new one is clear in intent.

**Review by jianoaix — APPROVED** [review]

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (clarkzinzow) [inline python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:100]: Nit: Eliminate an extra `self._create_task()` call path.
```suggestion
            if self._block_bundle:
                bundle, self._block_bundle = self._block_bundle, bundle
            self._create_task(bundle)
```

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (clarkzinzow) [inline python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:119]: Nit: Move `ref` deletion to right after it is no longer needed, to possibly free up space before `ray.get(all_refs[-1])` completes.
```suggestion
        all_refs = list(ray.get(ref))
        del ref
        block_refs = all_refs[:-1]
        block_metas = ray.get(all_refs[-1])
```

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (clarkzinzow) [inline python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:128]: Do we have an idea of how to address this TODO?

File: python/ray/data/_internal/execution/operators/map_operator.py
Comment (clarkzinzow) [reply python/ray/data/_internal/execution/operators/map_operator.py:46]: @ericl But that intent is incorrect: this is a target to get near to, not a minimum/floor. We add blocks to a bundle up to this target size, but we purposefully do not exceed it, so this is definitely not a minimum.

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (clarkzinzow) [inline python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:81]: Couldn't we eliminate this short-circuit by setting `self._min_rows_per_batch: int = op._min_rows_per_batch or 0` in the constructor? `get_num_rows` is not expensive, and we should always fall through to creating a task on the current bundle. With the suggestion that I have below, we'll then only have a single `self._create_task()` call in `add_input()`.

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (clarkzinzow) [inline python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:92]: This will drop empty blocks, right? We currently propagate empty blocks, IIRC.

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (clarkzinzow) [reply python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:170]: I think that's pretty easy to do with either executor: instead of looking at whether there is a 10x reduction in parallelism globally after block bundling, we keep a counter for the size of the of the **current** bundle, and if we've merged more than 10 bundles into a single bundle, we log a warning (if we haven't already logged a warning, i.e. log-once). Or we could keep a running average and log at the end of execution, although I think that the eager logging would be better.

That should work for both the bulk executor and the streaming executor, right?

File: python/ray/data/tests/test_operators.py
Comment (clarkzinzow) [inline python/ray/data/tests/test_operators.py:17]: Nit: Parentheses are redundant.
```suggestion
    for block, _ in bundle.blocks:
```

File: python/ray/data/_internal/execution/operators/map_operator.py
Comment (clarkzinzow) [inline python/ray/data/_internal/execution/operators/map_operator.py:107]: Most of this abstraction's API is a passthrough to the execution state abstraction, which is a bit of an abstraction hierarchy smell. I'm wondering if inverting the hierarchy or reworking the relationship into a compositional pattern would yield something cleaner.

I don't have any immediate ideas so maybe we can reexamine this later.

File: python/ray/data/_internal/execution/operators/map_operator.py
Comment (clarkzinzow) [inline python/ray/data/_internal/execution/operators/map_operator.py:58]: Another abstraction hierarchy smell: we have a circular dependency between the operator abstraction and its execution state, with each depending on each other.

**Review by clarkzinzow — COMMENTED** [review]
Mostly nits, the only potential blocker in my mind is the question around the block bundling logic: it appears to be dropping empty blocks, which I don't think is the current Datasets behavior.

File: python/ray/data/_internal/execution/operators/input_data_buffer.py
Comment (ericl) [reply python/ray/data/_internal/execution/operators/input_data_buffer.py:29]: It looks more concise, but less readable.

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (ericl) [reply python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:128]: Yeah, we have to add a "manual" ref count to the block bundle.

File: python/ray/data/_internal/execution/operators/map_operator.py
Comment (ericl) [reply python/ray/data/_internal/execution/operators/map_operator.py:48]: Alright, let me rename this to `min_rows_per_bundle` then. I don't think it's possible to be precisely unambiguous, and would prefer we keep the "min" intent which is the big picture.

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (ericl) [reply python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:163]: Let me add a TODO to do this later.

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (ericl) [reply python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:91]: Let me try removing this.

File: python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py
Comment (ericl) [inline python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py:82]: I ended up putting this back, in order to enable empty block propagation.

**Review by ericl — COMMENTED** [review]
Updated; main changes was I removed the circular dependency between the operator impl and the wrapper operator.

File: python/ray/data/_internal/execution/operators/input_data_buffer.py
Comment (clarkzinzow) [reply python/ray/data/_internal/execution/operators/input_data_buffer.py:29]: For sure, probably a toss up based on preference, keeping what you have is fine!

**Pull request merged**
